src/main/java/com/visenze/datatracking/DataTracking.java

```diff
@@ -0,0 +1,164 @@
+package com.visenze.datatracking;
+
+
+
+import android.content.Context;
+import android.util.Log;
+
+import com.visenze.datatracking.data.DataTrackingResponse;
+import com.visenze.datatracking.data.DeviceData;
+import com.visenze.datatracking.data.EventData;
+import com.visenze.datatracking.data.EventsBody;
+import com.visenze.datatracking.data.ResultData;
+import com.visenze.datatracking.utils.Utils;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.UUID;
+
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public class DataTracking {
+
+    public static final String TAG = "DataTracking";
+    public static final long SESSION_TIMEOUT = 900000; //15 min
+
+    private static DataTracking mInstance;
+
+    private APIService apiService;
+    private String code;
+    private String uid;
+    private long sid;
+
+    private List<EventData> events;
+    private DataCollection dataCollection;
+    private DeviceData mDeviceData;
+
+    private DataTracking() {
+        apiService = HttpInstance.getRetrofitInstance(Constants.BASE_URL).create(APIService.class);
+        events = new ArrayList<EventData>();
+    }
+
+
+    public static synchronized DataTracking getInstance() {
+        if(mInstance == null) {
+            mInstance = new DataTracking();
+
+        }
+        return mInstance;
+    }
+
+
+    public void init(Context context, String code) {
+        String uid = getUid(context);
+        this.code = code;
+        initDataTracking(context, uid);
+    }
+
+    public void init(Context context, String code, String uid) {
+        if(uid == null || uid.isEmpty()) {
+            uid = getUid(context);
+        }
+        this.code = code;
+        initDataTracking(context, uid);
+    }
+
+
+    private String getUid(Context context) {
+        String uid = Utils.getUid(context);
+        if(uid == null || uid.isEmpty()) {
+            uid = UUID.randomUUID().toString();
+            uid = uid.replace("-", "");
+            // save uid in preference.
+            Utils.saveUid(context, uid);
+        }
+        return uid;
+    }
+
+    private long getSid(Context context) {
+        long sid = Utils.getSid(context);
+        Date date = new Date();
```

Comment:
can refactor into another method isSessionExpired() ? Logically speaking, getSid is to retrieve, expire and generate new session should be separate logic. They should not be mixed together.

Reply:
pls also add unit test for this session management logic:
1) create new session
2) session expiration check
3) check for time zone e.g. user mess up the clock, will Date() still return the expected value?

Reply:
user should not need to pass in the Context again, see above. We do not need any UI element.

Reply:
it is private method. user will not pass the context here.
currently how sessionId will be generate is: 
when app launch it generate check the last session timestamp if it is within 30 minutes it will use the last timestamp. 
otherwise it just create a new session timestamp and overwrite in preference. 

i think it cover most of the case. let's say user launch app -> minimize it then change clock -> most likely the application will not be killed by android system. it will use the last sessionId. 
if user launch app -> kill app -> change clock > 30 minutes  -> relaunch app -> then new Date() of course will be using the system clock.which is the changed time. the only way to ensure use world clock is to get from API. Most game application use API to get world clock time.

Reply:
I understand your logic. As mentioned below under comments. We should separate the session ID and timestamp. If we use the timestamp as the session ID , then in the server log there would be a lot of common session IDs e.g. users are sending the events at exactly the same time. This will make it difficult to analyze and aggregate. Generate as a random UUID is better. Pls keep the timestamp separate.

"you can refer to Google sample session code: https://github.com/googleanalytics/autotrack/blob/master/lib/session.js#L153 Let's separate session id and timestamp. Session ID is supposed to be unique ideally. Can use the normal UUID function, no need to remove the "-" "

Reply:
as for the timestamp, I am not talking about user purposely screw up the time. I am talking about different time zone setting. As long as we can get the Unix timestamp i.e. number of milliseconds from 1970, then it is ok. This is independent from time zone. Even if the user changes the timezone, it should still be the same.

Reply:
added handling in session manager